### PR TITLE
try to avoid code duplication (buildErrorMsg and bindValidation)

### PR DESCRIPTION
**Pull request by polux**

### CODE REVIEW

File: src/Exec.hs [hunk 1]
```diff
@@ -152,22 +153,13 @@ runStep properties (ShellCmd workdir cmd mprop haltOnFailure) = do
   unless (null outmsg) $ putOut @s outmsg -- show step normal output, if any
   unless (null errmsg) $ putErr @s errmsg -- show step error output, if any
   let properties' = case mprop of Nothing -> properties
-                                  Just prop -> Map.insert prop (normalize outmsg) properties
+                                  Just prop -> Map.insert prop (trim outmsg) properties
```
Comment (smelc) [inline src/Exec.hs]: This is changing the behavior. We don't want to trim the property's value at the start nor do we want to eliminate trailing spaces. I'll restore the previous behavior which was doing the minimal possible change.

Comment (polux) [reply src/Exec.hs]: Ok, I'll change it back

Comment (polux) [reply src/Exec.hs]: Done

File: src/Exec.hs [hunk 2]
```diff
@@ -159,9 +158,11 @@ runStep properties (ShellCmd workdir cmd@Command{cmdString} mprop haltOnFailure)
   unless (null errmsg) $ putErr @s errmsg -- show step error output, if any
   unless (rc == ExitSuccess) $
     zzLogError @s (cmdString ++ " failed: " ++ show rc)
-  let properties' = properties & maybe id (`Map.insert` trim outmsg) mprop
+  let properties' = properties & maybe id (`Map.insert` normalize outmsg) mprop
   let streams = StepStreams (Just outmsg) (Just errmsg)
   return (properties', streams, toStatus rc, not haltOnFailure || rc == ExitSuccess)
+ where
+  normalize str = if last str == '\n' then init str else str
```
Comment (smelc) [inline src/Exec.hs]: How is the case where outmsg is empty covered?

Comment (polux) [reply src/Exec.hs]: Done

**Pull request merged**